lib/src/dbscan_base.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017, erikrahtjen. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+class DBSCAN<T> {
+  final List<T> _dataset;
+  final num _epsilon;
+  final num _minPts;
+  final Function _distance;
+
+  List<List<int>> _clusters = [];
+  List<int> _noise = [];
+  List<int> _visited = [];
+  List<int> _assigned = [];
+
+  DBSCAN(this._dataset, this._epsilon, this._minPts, this._distance(T a, T b)) {
+    _visited = new List(_dataset.length);
+    _assigned = new List(_dataset.length);
+
+    for (int pointId = 0; pointId < _dataset.length; pointId++) {
+      // if point is not visited, check if it forms a cluster
+      if (_visited[pointId] != 1) {
+        this._visited[pointId] = 1;
+
+        // if closest neighborhood is too small to form a cluster, mark as noise
+        var neighbors = _regionQuery(pointId);
+
+        if (neighbors.length < _minPts) {
+          _noise.add(pointId);
+        } else {
+          // create new cluster and add point
+          var clusterId = _clusters.length;
+          _clusters.add(new List());
+          _addToCluster(pointId, clusterId);
+
+          _expandCluster(clusterId, neighbors);
+        }
+      }
+    }
+  }
+
+  List<List<T>> get clusters {
+    List<List<T>> returnedClustersOfData = new List();
+
+    for (var cluster in _clusters) {
+      List<T> dataCluster = [];
+      cluster.forEach((int datumId) {
+        dataCluster.add(_dataset[datumId]);
+      });
+      returnedClustersOfData.add(dataCluster);
+    }
+
+    return returnedClustersOfData;
+  }
+
+  List<T> get noise {
+    List<T> returnedNoise = new List();
+
+    for (var datumId in _noise) {
+      returnedNoise.add(_dataset[datumId]);
+    }
+
+    return returnedNoise;
+  }
+
+  _regionQuery(int pointId) {
+    var neighbors = [];
+
+    for (var id = 0; id < _dataset.length; id++) {
+      var dist = _distance(_dataset[pointId], _dataset[id]);
+      if (dist < _epsilon) {
+        neighbors.add(id);
+      }
+    }
+
+    return neighbors;
+  }
+
+  void _addToCluster(int pointId, int clusterId) {
```

Comment:
I think that because the algorithm operates on top of lists of positions of items in a cluster, and take advantage of the properties of arrays, it'd be harder to manage than this, in the end. I am going to keep it as is. These method names come from the pseudocode presented with the algorithm